### DISCUSSION TITLE
Compilation with Clang?

### DISCUSSION THREAD

**Discussion by SharanSMenon** [rsta2/circle#305] [Q&A] [answered]
How can I compile circle with clang instead of arm-gcc?

I am on a M1 MacBook.

**Comment by rsta2** [comment, marked as answer]
Normally Clang is not supported officially, but I made an attempt to support it earlier on the branch *clang-support*. There are several limitations with this version, but if you want to try it, let me know, so I can I give you instructions on how to configure it.

**Reply by SharanSMenon** [reply]
I have cloned the `clang-support` branch and managed to build `libcircle.a`. However, when I try to build a sample with clang, I get the following error:

```
  CPP   main.o
  CPP   kernel.o
  LD    kernel8.elf
ld: unknown option: --Map
clang: error: linker command failed with exit code 1 (use -v to see invocation)
clang-13: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
make: *** [kernel8.img] Error 1
```
This same sample compiles just fine using the `arm-gcc` compiler. 

I am using clang from `brew install llvm` and not XCode clang.

**Reply by SharanSMenon** [reply]
I am using `AARCH=64` btw.

**Reply by rsta2** [reply]
You need a *Config.mk* file in the Circle root directory with a contents like this:

```
CLANG = 1
AARCH = 64
RASPPI = 3
```

If your linker still does not recognize the `--Map` option, you can try to remove `-Wl,--Map,$(TARGET).map` in [this line](https://github.com/rsta2/circle/blob/clang-support/Rules.mk#L206).

**Reply by SharanSMenon** [reply]
I built circle using the `Config.mk` that you showed me. When I try to built the sample without the `-Wl,--Map,$(TARGET).map`, I get this error now:

```
  LD    kernel8.elf
ld: unknown option: --section-start=.init=0x80000
clang: error: linker command failed with exit code 1 (use -v to see invocation)
clang-13: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
make: *** [kernel8.img] Error 14
```

**Reply by rsta2** [reply]
The linker, which is spawned by clang seems to be incompatible with the one, I have here. I tested the clang support on Linux only. It would require a longer session to get this running. Given that the clang support is not officially available, I'm in doubt, if it makes sense to continue here.

**Reply by SharanSMenon** [reply]
So does that mean that there are no options to compile this on an Apple Silicon Mac?

I don't think `arm-gcc` works on M1 Macs.

**Reply by rsta2** [reply]
Honestly I don't know. I do not have a Mac and are not familiar with it. Perhaps somebody else tried to build Circle on a Mac?

**Comment by rsta2** [comment]
I have merged the *clang-support* branch to the *develop* branch to be part of the next release. The branch *clang-support* has been removed. There are instructions on using Clang/LLVM to build Circle in the file *doc/clang-support.txt*. The problem with the linker on Macs is not solved so far. It would be great, if somebody could deal with this, who has a Mac, because I do not have one.
